File: java/src/leetcode/leetcode138.java

class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}


class Solution {
    public Node copyRandomList(Node head) {
        if(head == null){
            return null;
        }
        for(Node node = head;node != null;node = node.next.next){
            Node newnode = new Node(node.val);
            newnode.next = node.next;
            node.next = newnode;
        }
        for(Node node = head;node != null;node = node.next.next){
            Node newnode = node.next;
            if(node.random != null){
                newnode.random = node.random.next;
            }
        }
        Node newhead = head.next;
        for(Node node = head;node != null;node = node.next){
            Node newnode = node.next;
            node.next = node.next.next;
            newnode.next = (newnode.next != null) ? newnode.next.next : null;
        }
        return newhead;
    }
}

File: py/leetcode138.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        cur = head
        while cur:
            cur.next = Node(cur.val,cur.next)
            cur = cur.next.next
        cur = head
        while cur:
            if cur.random:
                cur.next.random = cur.random.next
            cur = cur.next.next
        cur = dummy = Node(0, head)
        while cur.next:
            cur.next = cur.next.next
            cur = cur.next
        return dummy.next